plugins/providers/virtualbox/action/prepare_clone_snapshot.rb

```diff
@@ -0,0 +1,65 @@
+require "log4r"
+
+require "digest/md5"
+
+module VagrantPlugins
+  module ProviderVirtualBox
+    module Action
+      class PrepareCloneSnapshot
+        def initialize(app, env)
+          @app = app
+          @logger = Log4r::Logger.new("vagrant::action::vm::prepare_clone")
+        end
+
+        def call(env)
+          if !env[:clone_id]
+            @logger.info("no clone master, not preparing clone snapshot")
+            return @app.call(env)
+          end
+
+          # If we're not doing a linked clone, snapshots don't matter
+          if !env[:machine].provider_config.linked_clone
+            return @app.call(env)
+          end
+
+          # We lock so that we don't snapshot in parallel
+          lock_key = Digest::MD5.hexdigest("#{env[:clone_id]}-snapshot")
+          env[:machine].env.lock(lock_key, retry: true) do
```

Comment:
As I've understood, `Environment#lock` provides only a process-level lock, so it will not work in parallel multi-machine environment (`vagrant up vm_1 vm_2`) due to this line
https://github.com/mitchellh/vagrant/blob/c17efbed755df621d6571742e163e8dff61dd119/lib/vagrant/environment.rb#L508

I guess there should be some global lock, like this:
https://github.com/mitchellh/vagrant/blob/b721eb62cfbfa93895d0d4cf019436ab6b1df05d/plugins/synced_folders/nfs/synced_folder.rb#L79